### DISCUSSION TITLE
Error with import.meta.env

### DISCUSSION THREAD

**Discussion by simplecommerce** [vikejs/vike#1068] [Discussion] [answered]
Hi, 

I am currently testing this library with our existing project that is in TS.

I am running into a weird issue where I always get the following error around a file:

```
1:55:20 PM [vite] Error when evaluating SSR module /imports/core/imports/functions/server/index.ts: failed to import "/imports/core/imports/functions/server/introspectionQuery.ts"
|- TypeError: code.replaceAll is not a function
    at /root/v3/core-vite-ts/backend/node_modules/vite-plugin-ssr/dist/cjs/node/plugin/plugins/envVars.js:61:29
    at Array.forEach (<anonymous>)
    at TransformContext.transform (/root/v3/core-vite-ts/backend/node_modules/vite-plugin-ssr/dist/cjs/node/plugin/plugins/envVars.js:38:18)
    at Object.transform (file:///root/v3/core-vite-ts/backend/node_modules/vite/dist/node/chunks/dep-df561101.js:44283:62)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async loadAndTransform (file:///root/v3/core-vite-ts/backend/node_modules/vite/dist/node/chunks/dep-df561101.js:54950:29)
    at async instantiateModule (file:///root/v3/core-vite-ts/backend/node_modules/vite/dist/node/chunks/dep-df561101.js:55875:10)
```

The code inside is simply this as a test to isolate the issue:

```
const myFunction = () => {
	const test = import.meta.env.PUBLIC_ENV__GRAPHQL_URL;

	console.log(test);
}
```

I have a .env file in my root project folder with that same variable declared.
If I opend the `envVars.js` and add a console.log with the varName and val, I do see my variable name and value.

I don't understand what the issue is as it doesn't really tell me much even with the debug mode on.

Any help/hint is appreciated, thanks!

**UPDATE**

Ok so I did a console.log of `console.log(import.meta.env)` and I see only the following variables:

```
{
  BASE_SERVER: '/',
  BASE_ASSETS: '/',
  BASE_URL: '/',
  MODE: 'development',
  DEV: true,
  PROD: false,
  SSR: true
}
```

so its not seeing my .env file or loading it for some reason.

**UPDATE 2**

That is odd, I had to add the following to my `vite.config.js`:

```
	envPrefix: "PUBLIC_ENV__",
```

In order to see my variables, I assumed that it would do it automatically based on the doc.

It might be my setup that is wrong.
Also what I find odd, is that the file being imported is only imported from my `_default.page.server.jsx` so I assumed it would see the env variable without having to make it public.

If I do 

```
console.log(import.meta.env)  // this works and shows me the PUBLIC_ENV__GRAPHQL_URL
console.log(import.meta.env.PUBLIC_ENV__GRAPHQL_URL) // this causes it to crash for some reason with the same error.
```

I'll try a project from scratch and try to fill it from there.

**UPDATE 3**

My code seems to be the issue, I tried to start from a simple example and I was able to get up to the point of seeing a hello world page.
Now what I seem to see is that in functions that I import from _default.page.server.jsx they do not see to have access to all import.meta.env but I do have access if I do process.env instead. I don't know if this is normal behavior or not.

Ok so further testing shows that when I do `console.log(import.meta.env)` I can see the `SSR` value to `true`.
But if I do conditional logic like:

```
if (import.meta.env.SSR) {
  // ... server only logic
}
```

It crashes with 

```
 TypeError: code.replaceAll is not a function
    at /root/v3/core-vite-ts/backend/node_modules/vite-plugin-ssr/dist/cjs/node/plugin/plugins/envVars.js:60:29
    at Array.forEach (<anonymous>)
    at TransformContext.transform (/root/v3/core-vite-ts/backend/node_modules/vite-plugin-ssr/dist/cjs/node/plugin/plugins/envVars.js:38:18)
    at Object.transform (file:///root/v3/core-vite-ts/backend/node_modules/vite/dist/node/chunks/dep-df561101.js:44283:62)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async loadAndTransform (file:///root/v3/core-vite-ts/backend/node_modules/vite/dist/node/chunks/dep-df561101.js:54950:29)
    at async instantiateModule (file:///root/v3/core-vite-ts/backend/node_modules/vite/dist/node/chunks/dep-df561101.js:55875:10)
```

**Comment by brillout** [comment]
What vite-plugin-ssr version are you using?

If you can reproduce the following with the latest version then the best would be a reproduction (it can but doesn't have to be minimal).

```
 TypeError: code.replaceAll is not a function
    at /root/v3/core-vite-ts/backend/node_modules/vite-plugin-ssr/dist/cjs/node/plugin/plugins/envVars.js:60:29
```

**Reply by simplecommerce** [reply]
I am using the following versions:

vite@4.4.9
typescript@5.1.6 
vite-plugin-ssr@0.4.135 
@vitejs/plugin-react@4.0.4
node v14.18.1
npm 6.14.15

I will try to see if I can reproduce it with a minimal repo

**Reply by simplecommerce** [reply]
Here is the repo: https://github.com/simplecommerce/vite-plugin-ssr-env-bug/tree/main

I did a minimal example and I kept my original vite.config.ts and tsconfig.json with package.json.

**Reply by simplecommerce** [reply]
I can also reproduce it using this: https://vite-plugin-ssr.com/scaffold and selecting react + ts.
And creating a `/imports/server.ts` and adding this code inside:

```
if (import.meta.env.SSR) {
    console.log("HELLO WORLD")
}
```
and then importing this file in the `_default.page.server.ts` file.

I'm assuming its my logic that is wrong and that I don't understand how the framework works properly.

**Reply by brillout** [reply]
I didn't manage to reproduce.

Your reproduction doesn't build:

```
npm ERR! code 128
npm ERR! An unknown git error occurred
npm ERR! command git --no-replace-objects ls-remote ssh://git@gitlab.com/prosomo/ckeditor5-sc-plugins.git
npm ERR! git@gitlab.com: Permission denied (publickey).
npm ERR! fatal: Could not read from remote repository.
npm ERR! 
npm ERR! Please make sure you have the correct access rights
npm ERR! and the repository exists.

npm ERR! A complete log of this run can be found in: /home/gitpod/.npm/_logs/2023-08-13T12_56_01_106Z-debug-0.log
```

I created another reproduction using your instructions: https://github.com/brillout/vps-env-replaceAll-reprod. But it seems to work just fine. Can you reproduce using my reproduction attempt?

**Reply by simplecommerce** [reply]
Here is a small video with the steps using your repo: https://www.loom.com/share/48c8119ab70d4fae977d2842eee30bac?sid=535d4551-fdd9-4bf6-945e-a4c1f7dcf3e8

**Reply by simplecommerce** [reply]
Hi, I just did a small test using another version of node, I couldn't use the latest because of my ubuntu setup but I tried v17.0.1 and it works fine.

I tried installing v14.21.3 and testing it again but it fails.

So for some reason it seems to be related to this.

Oh, and I am not looking to keep v14, I want to update to latest, I was just testing the framework.

**Reply by brillout** [reply, marked as answer]
VPS only supports Node.js >=16.0.0. I've added an assertion to improve the error message: 3b305cbb83a3a9667c5ce1565520aaf5dca28033

Let me know if you still run into issues with environment variables; there have been many changes recently so some things may still need polishing.

**Reply by simplecommerce** [reply]
Its fine so far, thanks for your help, I appreciate it!

**Reply by brillout** [reply]
👍 Btw. in case your company is up for it: https://github.com/sponsors/brillout.
